### ISSUE TITLE
Model wrapper for controlling iterative models.

### ISSUE THREAD

**Issue by ablaom** [alan-turing-institute/MLJ.jl#139]
I would like a common interface for implementing "early stopping" of iterative methods, with possibly an option to adapt a  "learning rate" or other hyperparameter as iteration progresses. 

To buy into this interface (ie to get performance benefits)  a model need only to implement the  MLJBase.update method, so that increases to the iteration parameter do not trigger retraining from scratch (which they should do anyway). Also, any parameter, such as a "learning rate" that is to be adapted, should similarly not trigger retraining from scratch. An existing model that does this already (without the extra adaptive parameter) is EnsembleModel. @ayush-1506 and I are currently arranging this for the MLJ Flux models we are working on. **Edit** Done.

We could implement this as a model wrapper, just like TunedModel.  **Edit:** Like the TunedModel, we would specify a resampling strategy from the ones that already exist for this purpose (eg, CV()), for specifying how a performance is estimated for purposes of deciding early stopping.

What do people think?

**Comment by darenasc** [comment]
Like in a general base wrapper for iterative methods? sounds good.

An alternative would be to pass an optional early_stopping parameter to the fit function and pass it using the update method?

**Comment by ablaom** [comment]
Thanks for that!

I'd rather avoid passing parameters to calls to fit!. The philosophy in such cases has been to have an "instruction"  hyperparameter for such models, but for then we need to ensure all model implementers of iterative methods use the same convention for naming this, and so forth. The advantage of my suggestion is that no such convention is necessary. All choices are "externalised" to the wrapper design.

We already have something like this in `learning_curve!`. You can generate a learning curve(s) with *any* choice of hyperparameter - it doesn't have to be the iteration parameter. If it is an "iteration parameter" (we never actually define this term and don't recally need do) then the function runs efficiently, assuming increasing that parameter (and leaving others alone) does not retrigger retraining from scratch (because there is a sensible overloading of the model update method).

I should add that if a sensible update method is not implemented, then we can get the same behaviour *without this knowledge*, it just takes longer. (Maybe should think of a way to expose this fact to the user?)

**Comment by vollmersj** [comment]
This sort of exists in [LearningStrategies.jl](https://github.com/JuliaML/LearningStrategies.jl).

**Comment by fkiraly** [comment]
I completely agree with @ablaom that "update" should be a full-blown interface point, rather than just having it within the interfaced models. 
The argument that meta-astrategies that make use of learning curves etc are externalized I think makes this a canonical choice. Most importantly, perhaps, it would allow the common "inspect learning curve of neural network and eyeball and then do a black magic dance while inspecting the flight of the birds and the learning curve's entrails until you have found the right learning parameters" part of the deep learning workflow more reproducible and accessible to systematic automation.

I also remember a discussion about "update" somewhere else (other issue), but can't point my finger to it, so plead add issue no. if anyone remembers.

Anyway, I thin we ought to think carefully about:
(i) use cases. I think there are *multiple* significant use cases for an "update" interface point.
(ii) the method signatures, e.g., what is a "parameter" of update and what is a parameter of the method.

(i) use cases

There's multiple I can think of:
- on-line learning. Supervised learning in essence, the step of "add more training data and update the model fit". Only needs the "update" interface point.
- active learning. Give the method the feature/label pairs it asked for. Needs to be designed with a "query" in mind that might be a callback (?), or maybe a completely different interface point?
- Bayesian update. The model is fully Bayesian, and we use the method's on-board inference engine to update parameter (posteriors). This needs "update", and model parameters which are distributions(.jl)
- "one more iteration". Many - but by far not all - models rely on some internal optimizagion (SGD etc). One could use "update" to tell the model to continue "running", without ingesting more data.

The key issue that I see here is that the use cases rely to different extent on:
- continuing with model computation, or
- ingesting data
I think it's a valid question whether one should make these *two* interface points rather than one. Or, whether one should allow for multiple "updaters" which would then become first-order types along the lines of "myupdater : modelupdaterfor{mysupervisedlearner}".
Or, one could require the "update" function to cover the case where no additional data is passed?

(ii) hyper-parameters and input of the update
I agree with @ablaom that update-related meta-strategies should externalize parameter choices in a wrapper that calls update of the wrapped object.

However, "update" still needs inputs: if you think about the key case of neural networks, you could choose epoch and batch size in reference to a larger data set.
I'd leave thse with the wrapper, which has these as parameters? The wrapped update then sees the "right sized" dataset, sub-sampled from the larger training set, presented by the wrapper?

Is that what you had in mind, @ablaom ?

**Comment by ablaom** [comment]
@fkiraly In principle I think all of these use cases could be accommodated by the existing design (with a small question mark next to online-learning). However, I would suggest that only the case of fixed data (the last use case) is within the scope of core functionality. The active learning and online learning cases are significantly more complex because they require some sort of communication from the wrapper to the core algorithm about what part of the data that has been bound to the machine should actually be accessed at any give stage in the "progressive" computation. So some sort of convention about this has to be established, and the model implementer (provider of the core algorithm) has to understand this convention, and so forth. 

@sjvollmer I think it is possible that LearningStrategies.jl could be used to save a little code, although the package itself is only about 400 lines. By using it we pay the price of slightly more abstraction and I'm not sure it's worth it.  Some experimenting is in order: Can you use LearningStrategies.jl with MLJ to train a random forest along the lines in examples/ using a "patience" strategy. This strategy (which is not provided out-of-the-box by LearningStrategies) is the one that says you stop adding trees when the loss on a holdout set increases `n` consecutive times. Compare the code you write to the code you would write to do the same "by hand".

**Comment by fkiraly** [comment]
@ablaom regarding online and active learning, I agree that this is an advanced interface question.

The main point that I was trying to make is, let's ensure interface back-compatibility, i.e., let's think about what "update" looks like so it's easy to extend to the more complex cases.
Both advanced cases will probably need an "update" interface point, plus sth else, and the former should probably be the same "update" interface point as in the simple.

The unresolved question I'd like to spell out is: should "one more iteration" be the same interface point as "one more data point" (or N more etc).

Regarding the advanced tasks, this is currently perhaps less importants, but have some random thoughts:
One way could be to wrap a data source in a model, e.g., a "data stream" data provider, or an "active learning" data provider, possibly with callbacks.

For online learning, we've had a discussion a while ago in #60, in relation to onlinestats, @joshday or @knutjaegersberg may want to comment. 
There's also two sub-cases, reactive to "appearance of data trigger" and reactive to "user trigger". For the latter, a specific data provider or data source type may not be necessary.

**Comment by ablaom** [comment]
@fkiraly 

`update` exists as a model method and is described in the documentation, along with how it fits in with the fitting of machines . See also the doc string for `fit!` below.  If the existing model API will not serve as an interface point for "one more data point"  (hopefully it will) then I don't think there's much we can do about it now. While we might add to the model API (in MLJBase) I am very reluctant to break anything there. 

You raise some good questions. I will address them best I can at #60 when I have time, thanks.

**Comment by ablaom** [comment]
fit!(mach::Machine; rows=nothing, verbosity=1, force=false)

When called for the first time, call `MLJBase.fit` on `mach.model` and
store the returned fit-result and report. Subsequent calls do nothing
unless: (i) `force=true`, or (ii) the specified `rows` are different
from those used the last time a fit-result was computed, or (iii)
`mach.model` has changed since the last time a fit-result was computed
(the machine is *stale*). In cases (i) or (ii) `MLJBase.fit` is 
called on `mach.model`. Otherwise, `MLJBase.update` is called.

**Comment by ablaom** [comment]
### Basic proposal for `IterativeModel` wrapper:


To be realistic, I think we restrict performance estimates to
ones based on a Holdout strategy. We don't currently have a way to do
"incrementally" perform cross-validation or similar.

The wrapper specifies a range of values for a specified iteration
parameter, just like for `learning_curve!`.

Fitting the wrapped model will progressively train the atomic model on
train set and at each stage estimate the performanc on a test
set. This will build a history of performance, to which various
criteria can be applied. Formally, a *criterion* is just a
boolean-valued function of the history.

At each stage in the training process we apply one or more *controls*
specified by the user. A *control* consists of one more criteria and
an *action*. An action is either "stop" or "change some parameter"
(e.g., a learning rate, or choice of optimizer). 

I'm not sure how much of this can be implemented by building upon
LearningStrategies.jl. Perhaps @ayushh-1506 can comment after some
experimentation?

A simpler design option would be just to provide a few simple options
for stopping.

*Side note:* To buy into this interface (that is, for this inteface to
actually work efficiently) the model being wrapped must overload the
MLJBase `update` method appropriately. See previous  comment for logic
governing `update` call by machine `fit!`. Specifically, the update
method should only force retraining from scratch (the fallback option)
if some hyperparameter other than the iteration parameter or some
"control" parameter, such as learning rate, has been changed since the
last call to `fit` or `update`. 

Formally, we have made no distinction between "control" parameters and
ordinary parameters. Since we have a safe fallback for `update`, which simply
treats all parameters as "non-control", I haven't been too concerned
with this.

How does this sound?

**Comment by fkiraly** [comment]
Sounds great, @ablaom .

In particular, I agree with:
- exposing "control" parameters via the same hyper-parameter interface points
- use of `fit!`as currently specified

I'd recommend however (thoughts?) that:
- abstract model performance evaluation methodology (simple hold-out, standard losses, machines/learners against data, nothing fancy) to be implemented first. 
Not sure whether that already exists? But it sounds like you would want to call that from within the model, rather than implementing from scratch. Just to have it as an interface point call that can be expanded, as opposed to custom code.
- a "control" object (struct type) is designed, with potential subcases being iteration-control, resample-control, evaluation-control, experiment-control, tuning-control, etc, not necessarily all to be implemented now.

**Comment by ablaom** [comment]
@fkiraly Thanks for the feedback.

> Not sure whether that already exists?

Errrr, I believe `evaluate!` already does what we want, no?

```julia
julia> forest.model.n = 20
20

julia> evaluate!(forest, verbosity=3, resampling=Holdout())
┌ Info: Evaluating using a holdout set. 
│ fraction_train=0.7 
│ shuffle=false 
│ measure=MLJ.rms 
│ operation=StatsBase.predict 
└ Resampling from all rows. 
[ Info: Training Machine{DeterministicEnsembleModel{DecisionTreeRegressor}} @ 1…09.
[ Info: One hash per new atom trained: 
####################
7.564343735989894

julia> forest.model.n = 23
23

julia> evaluate!(forest, verbosity=3, resampling=Holdout())
┌ Info: Evaluating using a holdout set. 
│ fraction_train=0.7 
│ shuffle=false 
│ measure=MLJ.rms 
│ operation=StatsBase.predict 
└ Resampling from all rows. 
[ Info: Updating Machine{DeterministicEnsembleModel{DecisionTreeRegressor}} @ 1…09.
[ Info: Building on existing ensemble of length 20
[ Info: One hash per new atom trained: 
###
7.757009385251871
```

```julia
julia>?evaluate!
```
    evaluate!(mach, resampling=CV(), measure=nothing, operation=predict, force=false, verbosity=1)

Estimate the performance of a machine `mach` using the specified
`resampling` strategy (defaulting to 6-fold cross-validation) and `measure`,
which can be a single measure or vector. 

Although evaluate! is mutating, `mach.model` and `mach.args` are
preserved.

Resampling and testing is based exclusively on data in `rows`, when
specified.

If no measure is specified, then `default_measure(mach.model)` is used, unless
this default is `nothing` and an error is thrown.

**Comment by ablaom** [comment]
> a "control" object (struct type) is designed, with potential subcases being iteration-control, resample-control, evaluation-control, experiment-control, tuning-control, etc, 

Don't forget, it should be able to cook toast too.

> Should use the command pattern, if it's admissible to use the term in Julia...

Thanks. If this translates into a Julia pattern and is appropriate, will use.

**Comment by fkiraly** [comment]
>Don't forget, it should be able to cook toast too.

Toast-control, indeed the best thing since sliced bread.

>Errrr, I believe evaluate! already does what we want, no?

Indeed it does!
What I was thinking about was a full benchmark orchestration interface, but I did miss the valid point that a model-local dispatch does the trick as well, and better.

**Comment by fkiraly** [comment]
But seriously, the "control" object discussion is perhaps worthwhile having.

It's a kind of input that appears in multiple places - and what I meant is not to have one object that does everything, but a common interface design.

**Comment by ablaom** [comment]
https://www.juliabloggers.com/julia-nomenclature/

**Comment by jpsamaroo** [comment]
I'm interested in adding active learning capabilities to MLJ so that I can train models from real-world streaming data like camera or microphone input. I assume this issue best fits that need? If so, can someone give me an example of what the `update` call to do active learning should look like from the user's perspective?

**Comment by ablaom** [comment]
Online/active learning is not currently supported by MLJ. The `update` method is generally for "more iterations" as opposed to "more data". Although this thread is obviously related, the more appropriate place to discuss this would be #60. 

I think this one is a little way off (months), as I should probably be the one to implement the core design and my priorities are currently elsewhere. There'll probably be an new `add_data` method for models and this will have to be implemented for each model you want to fit online. Obviously, as not every model can do this, there won't be a fallback.

**Comment by ablaom** [comment]
Some other things to keep in mind in this design: snapshots and function callbacks.

**Comment by ablaom** [comment]
Okay, I finally got around to firming up this design.

We now have two new packages for generic control of iterative models: [EarlyStopping.jl](https://github.com/ablaom/EarlyStopping.jl) and [IterationControl.jl](https://github.com/ablaom/EarlyStopping.jl).

My proposal for integrating them into MLJ is a wrapper as previously discussed in this thread. The following draft doc-string shows what I have in mind.

Feedback welcome.

cc  @lorenzoh @ToucheSir 

---

    IteratedModel(model=nothing,
                  controls=[Step(10), Patience(5), NumberLimit(50), NotANumber()],
                  final_train=false,
                  resampling=Holdout(),
                  measure=nothing,
                  weights=nothing,
                  class_weights=nothing,
                  operation=predict,
                  verbosity=1,
                  check_measure=true,
                  iteration_parameter=nothing,
                  cache=true)

Wrap the specified `model <: Supervised` in the specified `controls`.

Here `model` should support iteration (i.e.,
`iteration_parameter(model)` is different from `nothing`). Controls
are discussed further below.

To make out-of-sample losses available to the controls, `model` is
only trained on part of the data, as controlled iteration proceeds.
See details on training below. Specify `final_train=true` to ensure
the model is retrained on *all* available data, using the same number
of iterations, once controlled iteration has stopped.

Specify `resampling=nothing` if all data is to be used for controlled
iteration, with each out-of-sample loss replaced by the most recent
training loss, if reported by the model. Otherwise, `resampling` must
have type `Holdout`.

Assuming `final_train==true` or `resampling==nothing`,
`iterated_model` behaves exactly like the original `model` but with
the iteration parameter automatically selected. If
`final_train==false` (default) and `resampling!=nothing`, then
`iterated_model` behaves like the original model trained on a subset
of the provided data.

Controlled iteration can be continued (warm restart) by mutating a
control, or by mutating the iteration parameter of `model`, which is
otherwise ignored.

If the constructor cannot automatically determine the name of the
iteration parameter, then `iteration_parameter` will need to be
specified. This is a `Symbol`, or, in the case of a nested
hyper-parameter, an `Expr`, such as `:(booster.nrounds)`.


### Training

Given an instance `iterated_model` of `IteratedModel`, calling
`fit!(mach)` on a machine `mach = machine(iterated_model, data...)`
performs the following actions:

- The `data` is split into *train* and *test* sets, according to the
  specified `resampling` strategy, which must have type `Holdout` or
  be `nothing`.

- A clone of the wrapped model, `iterated_model.model`, is bound to
  the train data in an internal machine (called `train_mach`
  below). If `resampling === nothing`, all data is used instead. The
  iteration parameter of the clone is set to `1`.

- The specified `controls` are repeatedly applied to `train_mach` in
  sequence, until one of the controls triggers a stop. Loss-based
  controls (eg, `Patience()`, `GL()`, `Threshold(0.001)`) use an
  out-of-sample loss, obtained by applying `measure` to predictions
  and the test target values. (Specifically, these predictions
  are those returned by `operation(train_mach)`.)  If `resampling ===
  nothing` then the most recent training loss is used instead, which
  requires `supports_training_losses(model) === true`. Note that some
  controls require *both* out-of-sample and training losses (eg,
  `PQ()`).

- Once a stop has been triggered, a clone of `model` is bound to all
  `data` in a machine called `mach_production` below, unless
  `final_train==false` or `resampling==nothing`, in which case
  `mach_production` coincides with `train_mach`.


### Prediction

Calling `predict(mach, Xnew)` returns `predict(mach_production,
Xnew)`. Similar similar statements hold for `predict_mean`,
`predict_mode`, `predict_median`.


### Controls

The supported controls are:

- all controls provided by the package
  [`IterationControl.jl`](https://github.com/ablaom/IterationControl.jl#controls-provided)
  with the exception of `Data(...)`. The control wrappers
  `IterationControl.skip` (to apply a control only periodically) and
  `IterationControl.debug` are included.

- `Snapshot(; path="machine.jlso")`: Save the current state of
  `train_mach` (see above) to a file at `path` with a numerical
  annotation (as in, "machine_32.jlso").

A control is permitted to mutate the fields (hyper-parameters) of
`train_mach.model` (the clone of `model`). For example, to mutate a
learning rate one might use the control

    Callback(m -> m.model.eta = 1.05*m.model.eta)

However, unless `model` supports warm restarts with respect to changes
in that parameter, this will trigger retraining of `train_mach` from
scratch, with a different training outcome, which is not recommended.


### Warm restarts

If `iterated_model` is mutated and `fit!(mach)` is called again, then
a warm restart is attempted if the only parameters to change are
`model` or `controls` or both.

Specifically, `train_mach.model` is mutated to match the current value
of `iterated_model.model` and the iteration parameter of the latter is
updated to the last value used in the preceding `fit!(mach)` call. Then
repeated application of the (updated) controls begin anew.

**Comment by ablaom** [comment]
@jeremiedb @yalwan-iqvia

**Comment by ToucheSir** [comment]
I'll be pretty swamped for the next couple of weeks and won't have time to do this thread justice, but a couple of quick questions to get things rolling:

1. Does/how does this handle models like GANs that have different training code paths depending on the current iteration, loss trend, phase of the moon, etc?
2. Are hyperparams considered to be fixed upon model creation or can they be varied over time if someone is using e.g. https://github.com/darsnack/ParameterSchedulers.jl/?

**Comment by ablaom** [comment]
Closed as now implemented, but discussion to continue here.
